File: Python/785. IsGraphBipartite.py
'''
We are trying to check if a graph is bipartite.

A grpah is said to be bipartite if all the nodes of the graph can be
divided into two distinct sets such that there is no edge between 
the nodes belonging to the same set.

A graph with such a propety can be colored with only two colors.

Thus in this algorithm we will try to color the graph with two
colors and if we can the graph is bipartite else not.

'''

class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        colored = {}
        for i in range(len(graph)):
            if i not in colored and graph[i]:
                colored[i] = 1
                q = collections.deque([i])
                while q:
                    cur = q.popleft()
                    #checking the neighbors of current node
                    for j in graph[cur]:
                        if j not in colored:
                            colored[j] = 1-colored[cur]
                            q.append(j)
                        #neighbor is the same color as the current node
                        elif colored[j] == colored[cur]:
                            return False
        return True

File: C++/785.IsGraphBipartite.cpp
/*
   Using the following numbers for colors!
   white = -1, red = 1, blue = 0;
*/
class Solution {
public:
    //this function will now check the nodes adjacent to the current node(i): if its white, it will
    //change to blue and if its already filled, it will check the color of 2 connected nodes
    bool helper_dfs(int i, vector<int>& color, vector<vector<int>>& graph)
    {
        //checking for the nodes connected to ith node
        for(auto u: graph[i])
        {
            if(color[u]==-1)
            {
                //changing the color to blue if its white as we don't want adjacent nodes to have the
                //same color
                color[u] = 1 - color[i];
                //calling the helper function once again to check for the nodes connected to node u
                if(helper_dfs(u, color, graph)==false)
                {
                    return false;
                }
            }
            //if the 2 connected nodes have the same color, they are NOT bipartite
            else if(color[u]==color[i])
            {
                return false;
            }
        }
        return true;
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        //initializing a color vector filled with white originally
        vector<int> color(n, -1);

        //traversing over the graph and coloring nodes accordingly
        for(int i=0; i<n; i++)
        {
            if(color[i]==-1)
            {
                //changing the color to red if its white and then calling a helper funtion to check
                //for the adjacent nodes it is connected to
                color[i]=1;
                if(helper_dfs(i, color, graph)==false)
                {
                    //if 2 adjacent nodes have the same color, they are not bipartite as we cannot
                    //partition them into 2 independent sets, thus, return false
                    return false;
                }
            }
        }
        //if the colors are alternating, we can partition them, therefore, return true
        return true;
    }
};

/*
    TC: O(n), since each node of the graph is traversed exactly once
    SC: O(n), for color vector
*/
